File: modules/desktop/home/hyprland/eww/scripts/eww_bar_backend/display.py
import subprocess

from .common import parse_json, run_text
from .paths import display_mode_path
from .inhibitors import (
    idle_inhibited_state,
    lid_inhibited_state,
    set_idle_inhibited,
    set_lid_inhibited,
)


DISPLAY_MODES = {"normal", "external", "headless"}
INTERNAL_MONITOR_PREFIXES = ("eDP-", "LVDS-")

MODE_STATUS = {
    "normal": "Normal desktop mode",
    "external": "External display mode",
    "headless": "Headless server mode",
}


def read_display_mode():
    try:
        mode = display_mode_path().read_text().strip()
    except Exception:
        return "normal"
    return mode if mode in DISPLAY_MODES else "normal"


def write_display_mode(mode):
    path = display_mode_path()
    if mode == "normal":
        try:
            path.unlink()
        except FileNotFoundError:
            pass
        except Exception:
            pass
        return
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(mode)


def display_state(status=""):
    mode = read_display_mode()
    return {
        "mode": mode,
        "lid_inhibited": lid_inhibited_state(),
        "status": status or MODE_STATUS[mode],
        "class": mode,
    }


def monitor_state():
    return parse_json(run_text(["hyprctl", "monitors", "-j"]), [])


def is_internal_monitor(name):
    return name.startswith(INTERNAL_MONITOR_PREFIXES)


def split_monitors(monitors):
    enabled = [monitor for monitor in monitors if not monitor.get("disabled", False)]
    internal = [monitor for monitor in enabled if is_internal_monitor(monitor.get("name", ""))]
    external = [monitor for monitor in enabled if not is_internal_monitor(monitor.get("name", ""))]
    return internal, external


def run_hyprctl(*args):
    result = subprocess.run(
        ["hyprctl", *args],
        stdin=subprocess.DEVNULL,
        stdout=subprocess.DEVNULL,
        stderr=subprocess.DEVNULL,
        check=False,
    )
    if result.returncode != 0:
        raise RuntimeError(f"hyprctl {' '.join(args)} failed")


def set_display_mode(action):
    if action == "status":
        return display_state()

    if action == "toggle":
        # Toggle headless on/off. This is the wake path when the screen is off:
        # the popup "Turn screens on" button is unreachable in headless mode
        # (it renders on the display that was powered off), so a keybind bound
        # to this action lets the user press once to sleep and again to cancel.
        if read_display_mode() == "headless":
            return set_display_mode("normal")
        return set_display_mode("headless")

    if action == "restore":
        run_hyprctl("dispatch", "dpms", "on")
        return display_state("Screens turned on")

    if action == "normal":
        run_hyprctl("dispatch", "dpms", "on")
        run_hyprctl("reload")
        set_lid_inhibited(False)
        set_idle_inhibited(False)
        write_display_mode("normal")
        return display_state("Exited display mode")

    if action == "external":
        internal, external = split_monitors(monitor_state())
        if not external:
            raise ValueError("external mode requires an active external monitor")
        set_idle_inhibited(True)
        set_lid_inhibited(True)
        # Guarantee the external monitor is powered on: an "enabled" monitor can
        # still be DPMS-off (e.g. arriving from headless or an idle blank), in
        # which case disabling the internal panel would leave a black screen.
        run_hyprctl("dispatch", "dpms", "on")
        for monitor in internal:
            name = monitor.get("name", "")
            if name:
                run_hyprctl("keyword", "monitor", f"{name},disable")
        write_display_mode("external")
        return display_state("External monitors active")

    if action == "headless":
        set_idle_inhibited(True)
        set_lid_inhibited(True)
        run_hyprctl("dispatch", "dpms", "off")
        write_display_mode("headless")
        return display_state("Headless server mode")

    raise ValueError("display action must be normal, external, headless, restore, toggle, or status")

File: modules/desktop/home/hyprland/eww/go/internal/collect/display.go
package collect

import (
	"errors"
	"fmt"
	"time"

	"ewwbar/internal/paths"
)

// Display mode: which monitors are on, and whether the lid and idle inhibitors are
// held. The mode is persisted in a runtime file, so it survives an eww restart but
// not a reboot.

type Display struct {
	Mode         string `json:"mode"`
	LidInhibited string `json:"lid_inhibited"`
	Status       string `json:"status"`
	Class        string `json:"class"`
}

var displayModes = map[string]bool{"normal": true, "external": true, "headless": true}

var modeStatus = map[string]string{
	"normal":   "Normal desktop mode",
	"external": "External display mode",
	"headless": "Headless server mode",
}

// hyprctlTimeout, like systemctlTimeout, exists because these run on a
// control-socket worker with a bounded pool.
const hyprctlTimeout = 5 * time.Second

// ReadDisplayMode treats any unreadable file or unrecognised contents as "normal",
// so a corrupt runtime file cannot wedge the bar in a mode with no monitors on.
func ReadDisplayMode() string {
	text, ok := ReadTextFile(paths.DisplayMode())
	if !ok {
		return "normal"
	}
	if mode := Strip(text); displayModes[mode] {
		return mode
	}
	return "normal"
}

// WriteDisplayMode: "normal" is represented by the file's ABSENCE, not by its
// contents, which is why this deletes rather than writes. Both failure paths are
// swallowed, because raising would abort a switch that has already physically
// happened.
func WriteDisplayMode(mode string) {
	if mode == "normal" {
		_ = RemoveFile(paths.DisplayMode())
		return
	}
	_ = WriteTextFile(paths.DisplayMode(), mode)
}

// DisplayState: an empty status takes the mode's own description.
func DisplayState(status string) Display {
	mode := ReadDisplayMode()
	if status == "" {
		status = modeStatus[mode]
	}
	return Display{
		Mode:         mode,
		LidInhibited: LidInhibitedState(),
		Status:       status,
		Class:        mode,
	}
}

// MonitorState keeps only the JSON objects, which the original did not: a report
// that is not a list of objects raises partway through a display switch. The guard
// cannot simply be "is it a list" -- an EMPTY object survives the original
// untouched, because iterating it yields nothing.
func MonitorState() []map[string]any {
	var monitors []any
	if !ParseJSON(RunText(defaultTimeout, "hyprctl", "monitors", "-j"), &monitors) {
		return nil
	}
	out := []map[string]any{}
	for _, raw := range monitors {
		if monitor, isMap := raw.(map[string]any); isMap {
			out = append(out, monitor)
		}
	}
	return out
}

func runHyprctl(args ...string) error {
	if code := RunStatus(hyprctlTimeout, "hyprctl", args...); code != 0 {
		return fmt.Errorf("hyprctl %s failed", joinArgs(args))
	}
	return nil
}

// Hyprland 0.56 moved hyprctl's control surface to Lua, and the positional forms
// this file used are now syntax errors inside the generated call:
//
//	$ hyprctl dispatch dpms off
//	error: [string "return hl.dispatch(dpms off)"]:1: ')' expected near 'off'
//	exit 7
//
// The `keyword` form was worse, because it fails with exit 0:
//
//	$ hyprctl keyword general:border_size 1
//	keyword can't work with non-legacy parsers. Use eval.
//	exit 0
//
// runHyprctl sees only the code, so external mode would have reported success and
// claimed "External monitors active" with the internal panel still lit. The eval
// form errors with exit 7, which is what makes checking the code meaningful.
const (
	hyprDpmsOn  = `hl.dsp.dpms({ action = "on" })`
	hyprDpmsOff = `hl.dsp.dpms({ action = "off" })`
)

// hyprDisableMonitor is the Lua replacement for `keyword monitor <name>,disable`.
// The field is `disabled`; `disable` and `enabled` are both rejected as unknown.
//
// A partial rule RESETS the fields it omits -- a rule carrying only output and
// disabled reset a live monitor's scale from 1.0 to a default 1.5. Survivable only
// because "normal" runs `hyprctl reload` first, which re-reads the config and the
// per-output overrides in ~/.config/hypr/monitors.lua. Do not optimise that away.
func hyprDisableMonitor(name string) string {
	return fmt.Sprintf("hl.monitor{ output = %q, disabled = true }", name)
}

func joinArgs(args []string) string {
	out := ""
	for i, arg := range args {
		if i > 0 {
			out += " "
		}
		out += arg
	}
	return out
}

// ErrDisplayAction is display.set_display_mode's ValueError for an unknown
// action. Named so the control handler can tell a bad request from a failure.
var ErrDisplayAction = errors.New(
	"display action must be normal, external, headless, restore, toggle, or status")

// ErrNoExternalMonitor is the other ValueError: external mode with nothing to
// switch to would blank every screen.
var ErrNoExternalMonitor = errors.New("external mode requires an active external monitor")

// SetDisplayMode's ORDER of side effects is the contract, not an implementation
// detail:
//
//   - external turns DPMS on BEFORE disabling the internal panel. An "enabled"
//     monitor can still be DPMS-off, and disabling the panel first would leave a
//     black screen with no way back.
//   - normal turns DPMS on and reloads BEFORE dropping the inhibitors.
//   - every path writes the mode file LAST, so a failure part-way leaves the file
//     describing the mode the machine is actually in.
func SetDisplayMode(action string) (Display, error) {
	switch action {
	case "status":
		return DisplayState(""), nil

	case "toggle":
		// Toggle headless on and off. This is the wake path when the screen is
		// already off: the popup's "Turn screens on" button renders on the display
		// that was powered down, so a keybind lets one press sleep and the next cancel.
		if ReadDisplayMode() == "headless" {
			return SetDisplayMode("normal")
		}
		return SetDisplayMode("headless")

	case "restore":
		if err := runHyprctl("dispatch", hyprDpmsOn); err != nil {
			return Display{}, err
		}
		return DisplayState("Screens turned on"), nil

	case "normal":
		if err := runHyprctl("dispatch", hyprDpmsOn); err != nil {
			return Display{}, err
		}
		if err := runHyprctl("reload"); err != nil {
			return Display{}, err
		}
		if _, err := SetLidInhibited(false); err != nil {
			return Display{}, err
		}
		if _, err := SetIdleInhibited(false); err != nil {
			return Display{}, err
		}
		WriteDisplayMode("normal")
		return DisplayState("Exited display mode"), nil

	case "external":
		internal, external := SplitMonitors(MonitorState())
		if len(external) == 0 {
			return Display{}, ErrNoExternalMonitor
		}
		if _, err := SetIdleInhibited(true); err != nil {
			return Display{}, err
		}
		if _, err := SetLidInhibited(true); err != nil {
			return Display{}, err
		}
		if err := runHyprctl("dispatch", hyprDpmsOn); err != nil {
			return Display{}, err
		}
		for _, monitor := range internal {
			name, _ := monitor["name"].(string)
			if name == "" {
				continue
			}
			if err := runHyprctl("eval", hyprDisableMonitor(name)); err != nil {
				return Display{}, err
			}
		}
		WriteDisplayMode("external")
		return DisplayState("External monitors active"), nil

	case "headless":
		if _, err := SetIdleInhibited(true); err != nil {
			return Display{}, err
		}
		if _, err := SetLidInhibited(true); err != nil {
			return Display{}, err
		}
		if err := runHyprctl("dispatch", hyprDpmsOff); err != nil {
			return Display{}, err
		}
		WriteDisplayMode("headless")
		return DisplayState("Headless server mode"), nil
	}
	return Display{}, ErrDisplayAction
}
